megatron/tokenizer/tokenizer.py

```diff
@@ -68,14 +68,22 @@ def build_tokenizer(args):
 
 
 def _vocab_size_with_padding(orig_vocab_size, args):
-    """Pad vocab size so it is divisible by model parallel size and
-    still having GPU friendly size."""
+    """Apply the requested rules to change the size of the vocabulary"""
 
     after = orig_vocab_size
-    multiple = args.make_vocab_size_divisible_by * \
-        args.tensor_model_parallel_size
-    while (after % multiple) != 0:
-        after += 1
+    if args.pad_vocab_size_to is not None:
+        after = args.pad_vocab_size_to
+        if after < orig_vocab_size:
+            raise ValueError(
+                f"You asked to pad the vocabulary to {after} when the initial vocabulary size is "
+                f"{args.pad_vocab_size_to}. You can only pad to a higher value."
+            )
+    else:
+        # Pad vocab size so it is divisible by model parallel size and still having GPU friendly size.
+        multiple = args.make_vocab_size_divisible_by * \
+            args.tensor_model_parallel_size
+        while (after % multiple) != 0:
+            after += 1
```

Comment:
```suggestion
    if args.pad_vocab_size_to is not None:
        if args.pad_vocab_size_to  < orig_vocab_size:
            raise ValueError(
                f"You asked to pad the vocabulary to {args.pad_vocab_size_to} when the initial vocabulary size is "
                f"{orig_vocab_size}. You can only pad to a higher value."
            )
        after = args.pad_vocab_size_to
    else:
        # Pad vocab size so it is divisible by model parallel size and still having GPU friendly size.
        after = orig_vocab_size
        multiple = args.make_vocab_size_divisible_by * \
            args.tensor_model_parallel_size
        while (after % multiple) != 0:
            after += 1
```
You need to remove `after = orig_vocab_size`

Also though I know that you check for multiple of TP down the line, I think it'd make sense to have the check here.

Also I'd be in favor of not ignoring `make_vocab_divisible_by`. Just check that it works, if it doesn't then we raise an exception.